ipld/retriever.go

```diff
@@ -0,0 +1,178 @@
+
+	ses := r.newSession(ctx, dah)
+	for _, q := range qs {
+		// TODO: ErrByzantineRow/Col
+		eds, err := ses.retrieve(ctx, q)
+		if err == nil {
+			return eds, nil
+		}
+		// retry quadrants until we have something or error
+	}
+
+	return nil, format.ErrNotFound
+}
+
+type retrieverSession struct {
+	dag   format.NodeGetter
+	adder *NmtNodeAdder
+
+	treeFn rsmt2d.TreeConstructorFn
+	codec  rsmt2d.Codec
+
+	dah    *da.DataAvailabilityHeader
+	square [][]byte
+}
+
+func (r *Retriever) newSession(ctx context.Context, dah *da.DataAvailabilityHeader) *retrieverSession {
+	size := len(dah.RowsRoots)
+	adder := NewNmtNodeAdder(ctx, format.NewBatch(ctx, r.dag, format.MaxSizeBatchOption(batchSize(size))))
+	return &retrieverSession{
+		dag:   merkledag.NewSession(ctx, r.dag),
+		adder: adder,
+		treeFn: func() rsmt2d.Tree {
+			tree := wrapper.NewErasuredNamespacedMerkleTree(uint64(size)/2, nmt.NodeVisitor(adder.Visit))
+			return &tree
+		},
+		codec:  r.codec,
+		dah:    dah,
+		square: make([][]byte, size*size),
+	}
+}
+
+func (rs *retrieverSession) retrieve(ctx context.Context, q *quadrant) (*rsmt2d.ExtendedDataSquare, error) {
+	defer func() {
+		// all shares which were requested or repaired are written to disk with the commit
+		// we store *all*, so they are served to the network, including incorrectly committed data(BEFP case),
+		// so that network can check BEFP
+		if err := rs.adder.Commit(); err != nil {
+			log.Errorw("committing DAG", "err", err)
+		}
+	}()
+	// request and fill quadrant's into
+	// we don't care about the errors here, just need to request as much data as we can to be able to reconstruct below
+	rs.request(ctx, q)
+
+	// try repair
+	err := rsmt2d.RepairExtendedDataSquare(rs.dah.RowsRoots, rs.dah.ColumnRoots, rs.square, rs.codec, rs.treeFn)
+	if err != nil {
+		log.Errorw("not enough shares sampled to recover, retrying...", "err", err)
+		return nil, err
+	}
+
+	return rsmt2d.ImportExtendedDataSquare(rs.square, rs.codec, rs.treeFn)
+}
+
+type quadrant struct {
+	roots []cid.Cid
+	// Coordinates of each quadrant(x;y)
+	// ------  -------
+	// |  Q0 | | Q1  |
+	// |(0:0)| |(1:0)|
+	// ------  -------
+	// |  Q2 | | Q3  |
+	// |(0;1)| |(1;1)|
+	// ------  ------
+	x, y int
+}
+
+func (rs *retrieverSession) request(ctx context.Context, q *quadrant) {
+	ctx, cancel := context.WithTimeout(ctx, RetrieveQuadrantTimeout)
+	defer cancel()
+
+	size := len(q.roots)
+	for i, root := range q.roots {
+		nd, err := rs.dag.Get(ctx, root)
+		if err != nil {
+			log.Errorw("getting root", "err", err)
+			continue // so that we still try for other roots
+		}
+
+		// TODO(@Wondertan): GetLeaves should return everything it was able to request even on error,
+		// 	so that we fill as much data as possible
+		// get leaves of left or right subtree
+		nds, err := GetLeaves(ctx, rs.dag, nd.Links()[q.x].Cid, make([]format.Node, 0, size))
+		if err != nil {
+			log.Errorw("getting all the leaves", "err", err)
```

Comment:
```suggestion
			log.Errorw("failed retrieving all leaves", "err", err)
```

Reply:
The reason why the log string is “getting all the leaves” is because the full log line can be read like ERROR getting all the leaves. 

I saw how PL folks were doing this and followed this pattern. We have it already throughout the code.